lib/tests/misc.nix

```diff
@@ -2461,4 +2464,109 @@ runTests {
     };
     expected = "c";
   };
+
+  testMapAttrRecursiveDoc = {
+    expr = mapAttrsRecursive (path: value: concatStringsSep "-" (path ++ [value])) {
+      n = {
+        a = "A";
+        m = {
+          b = "B";
+          c = "C";
+        };
+      };
+      d = "D";
+    };
+
+    expected = {
+      n = {
+        a = "n-a-A";
+        m = {
+          b = "n-m-b-B";
+          c = "n-m-c-C";
+        };
+      };
+      d = "d-D";
+    };
+  };
+
+  testMapAttrRecursiveCondTraverseAll = {
+    expr =
+      let
+        attr = {
+          a = 0;
+          b = 1;
+          c = [ 0 1 ];
+          d = [ { a = 0; } ];
+          e = { a = 0; };
+        };
+      in
+        mapAttrsRecursiveCond (x: true) (path: x: path) (
+          attr
+          // { nestedAttr = attr; }
+          // { nestedList = [ attr attr ]; }
+        );
+    expected = {
+      a = [ "a" ];
+      b = [ "b" ];
+      c = [ "c" ];
+      d = [ "d" ];
+      e = { a = [ "e" "a" ]; };
+      nestedAttr = {
+        a = [ "nestedAttr" "a" ];
+        b = [ "nestedAttr" "b" ];
+        c = [ "nestedAttr" "c" ];
+        d = [ "nestedAttr" "d" ];
+        e = { a = [ "nestedAttr" "e" "a" ]; };
+      };
+      nestedList = [ "nestedList" ];
+    };
+  };
+
+  testMapDataRecursiveCondTraverseAll = {
+    expr =
+      let
+        attr = {
+          a = 0;
+          b = 1;
+          c = [ 0 1 ];
+          d = [ { a = 0; } ];
+          e = { a = 0; };
+        };
+      in
+        mapDataRecursiveCond (x: true) (path: x: path) (
+          attr
+          // { nestedAttr = attr; }
+          // { nestedList = [ attr attr ]; }
+        );
+    expected = {
+      a = [ "a" ];
+      b = [ "b" ];
+      c = [ [ "c" 0 ] [ "c" 1 ] ];
+      d = [ { a = [ "d" 0 "a" ]; } ];
+      e = { a = [ "e" "a" ]; };
+      nestedAttr = {
+        a = [ "nestedAttr" "a" ];
+        b = [ "nestedAttr" "b" ];
+        c = [ [ "nestedAttr" "c" 0 ] [ "nestedAttr" "c" 1 ] ];
+        d = [ { a = [ "nestedAttr" "d" 0 "a" ]; } ];
+        e = { a = [ "nestedAttr" "e" "a" ]; };
+      };
+      nestedList = [
+        {
+          a = [ "nestedList" 0 "a" ];
+          b = [ "nestedList" 0 "b" ];
+          c = [ [ "nestedList" 0 "c" 0 ] [ "nestedList" 0 "c" 1 ] ];
+          d = [ { a = [ "nestedList" 0 "d" 0 "a" ]; } ];
+          e = { a = [ "nestedList" 0 "e" "a" ]; };
+        }
+        {
+          a = [ "nestedList" 1 "a" ];
+          b = [ "nestedList" 1 "b" ];
+          c = [ [ "nestedList" 1 "c" 0 ] [ "nestedList" 1 "c" 1 ] ];
+          d = [ { a = [ "nestedList" 1 "d" 0 "a" ]; } ];
+          e = { a = [ "nestedList" 1 "e" "a" ]; };
+        }
+      ];
+    };
+  };
```

Comment:
The `false` condition is not covered. Please add test cases where the predicate returns both booleans in a way that shows the intended behavior.

Reply:
Indeed, I added a test for that.